feat(Unit Library): Allow archiving units

geoffreykwan opened the pull request:
@breity Can you take a look at the styling and make changes if necessary?

## Changes
- Teachers can archive units in the My Units section
- Added Archive/Restore option in the unit info popup 3 dots menu
- Added checkbox to each unit card
- Added global checkbox to select all units in view
- Added global Archive/Restore button to archive/restore multiple units at once
- Added drop down to switch between Active and Archived view

## Test
- Test with https://github.com/WISE-Community/WISE-API/pull/248
- Make sure you can archive/restore a single unit using the unit info popup 3 dots menu
- Make sure you can archive/restore multiple units at once
- Make sure you can switch between Active and Archived view
- When you select multiple units, make sure the count of the selected units is displayed
- If units are selected and you switch between Active/Archived view or go to the next page of units, all the units become unselected

Closes #1549

geoffreykwan committed:
feat(Unit Library): Allow archiving a unit from the unit card menu

#1549

geoffreykwan committed:
feat(Unit Library): Allow archiving multiple units at once

#1549

hirokiterashima commented on src/app/modules/library/select-all-projects-checkbox/select-all-projects-checkbox.component.ts:
```diff
@@ -0,0 +1,25 @@
+      this.selectedAllProjects = true;
+    }
+    this.selectAllProjectsChanged.emit(this.selectedAllProjects);
+  }
```
Do we need to set the class fields in the function, or can we rely on the ```@Input``` to set them correctly? Writing like this seems to work for me.
```suggestion
  protected selectAllProjectsCheckboxClicked(): void {
    this.selectAllProjectsChanged.emit(!(this.selectedAllProjects || this.selectedSomeProjects));  
  }
```

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.ts:
```diff
@@ -80,6 +117,69 @@ export class PersonalLibraryComponent extends LibraryComponent {
+    this.selectAllProjectsChanged(false);
+  }
+
+  protected selectAllProjectsChanged(selectAll: any): void {
```
Make function name active (instead of reactive, "...Changed"), and add boolean type?
```suggestion
  protected selectAllProjects(selectAll: boolean): void {
```

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.ts:
```diff
@@ -80,6 +117,69 @@ export class PersonalLibraryComponent extends LibraryComponent {
+
+  private clearSelectedProjects(): void {
+    this.selectAllProjectsChanged(false);
+  }
```
Do we need this function, or can we directly call ```selectAllProjects(false);```?
```suggestion
```

geoffreykwan replied:
The function makes the code clearer and reduces duplication. I've also renamed the function to unselectAllProjects() to make it clearer.

hirokiterashima commented on src/app/modules/library/library-project-menu/library-project-menu.component.ts:
```diff
@@ -38,6 +41,7 @@ export class LibraryProjectMenuComponent {
     this.isCanShare = this.isOwner() && !this.isRun;
     this.editLink = `${this.configService.getContextPath()}/teacher/edit/unit/${this.project.id}`;
     this.isChild = this.project.isChild();
+    this.archived = this.project.tags.includes('archived');
```
Encapsulate ```Project.tags``` field by defining ```Project.hasTag()```?
```suggestion
    this.archived = this.project.hasTag('archived');
```

hirokiterashima commented on src/app/services/archive-project.service.ts:
```diff
@@ -37,4 +38,74 @@ export class ArchiveProjectService {
+      project.tags.push('archived');
+    } else {
+      project.tags.splice(project.tags.indexOf('archived'), 1);
+    }
```
Should this be handled by a new ```archive()``` function that we define in ```Project```? Doing this would allow us to encapsulate the ```project.tags``` field.
```suggestion (pseudo-code)
    project.archive(archive);
```

hirokiterashima commented on src/app/services/archive-project.service.ts:
```diff
@@ -37,4 +38,74 @@ export class ArchiveProjectService {
+    this.refreshProjects();
+  }
+
+  updateProjectsArchivedStatus(projects: Project[], archived: boolean): void {
+    projects.forEach((project) => {
+      this.updateProjectArchivedStatus(project, archived);
+    });
+  }
+
+  showArchiveProjectsSuccessMessage(projects: Project[], archived: boolean): void {
+    this.snackBar
+      .open(
+        archived
+          ? $localize`Successfully archived ${projects.length} unit(s).`
+          : $localize`Successfully restored ${projects.length} unit(s).`,
+        $localize`Undo`
+      )
+      .onAction()
+      .subscribe(() => {
+        this.undoArchiveProjects(projects, archived ? 'unarchiveProjects' : 'archiveProjects');
+      });
+  }
+
+  undoArchiveProjects(projects: Project[], archiveFunctionName: string): void {
+    this[archiveFunctionName](projects).subscribe({
+      next: (response: ArchiveProjectResponse[]) => {
+        for (const projectResponse of response) {
+          const project = projects.find((project) => project.id === projectResponse.id);
+          this.updateProjectArchivedStatus(project, projectResponse.archived);
+        }
+        this.snackBar.open($localize`Action undone.`);
+      },
+      error: () => {
+        this.snackBar.open($localize`Error undoing action.`);
+      }
+    });
+  }
```
Should these functions be defined in PersonalLibraryComponent? It doesn't look like they're being used anywhere else?

geoffreykwan replied:
I copied this code from the archive runs code and put it into the ArchiveProjectService without deleting the code in the archive runs. I plan to refactor the code in another issue to make the archive runs and archive units both use the ArchiveProjectService.

hirokiterashima commented on src/app/modules/library/select-all-projects-checkbox/select-all-projects-checkbox.component.html:
```diff
@@ -0,0 +1,8 @@
+  [indeterminate]="selectedSomeProjects"
+  (click)="selectAllProjectsCheckboxClicked()"
+  aria-label="Select all projects"
+  i18n-aria-label
```
Do we want a tooltip instead of aria-label? We currently show a "Select" tooltip for the select runs checkbox.

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.ts:
```diff
@@ -1,21 +1,48 @@
+  protected showArchived: boolean = false;
 
-  constructor(protected dialog: MatDialog, protected libraryService: LibraryService) {
+  protected numSelectedProjects: Signal<number> = computed(() => {
+    return this.selectedProjects().length;
+  });
+  protected projectIdToIsSelected: Signal<any> = computed(() => {
+    return this.selectedProjects().reduce((accumulator, project) => {
+      accumulator[project.id] = true;
+      return accumulator;
+    }, {});
+  });
+  protected selectedAllProjects: Signal<boolean> = computed(() => {
+    const numSelectedProjects = this.numSelectedProjects();
+    return numSelectedProjects !== 0 && numSelectedProjects === this.getProjectsInView().length;
+  });
+  protected selectedSomeProjects: Signal<boolean> = computed(() => {
+    return this.numSelectedProjects() !== 0 && !this.selectedAllProjects();
+  });
```
For some of these signals, you can remove the inner bracket and the ```return```. For example:
```
  protected numSelectedProjects: Signal<number> = computed(() => 
    this.selectedProjects().length;
  );
```

hirokiterashima commented on src/app/modules/library/select-all-projects-checkbox/select-all-projects-checkbox.component.ts:
```diff
@@ -0,0 +1,25 @@
+export class SelectAllProjectsCheckboxComponent {
+  @Input() selectedAllProjects: boolean = false;
+  @Input() selectedSomeProjects: boolean = false;
+  @Output()
+  selectAllProjectsChanged: EventEmitter<boolean> = new EventEmitter<boolean>();
```
Put on one line?
Also, should this field be named ```selectAllProjectsEvent```, following Angular convention (https://angular.io/guide/inputs-outputs#configuring-the-child-component-1)?

hirokiterashima commented on src/app/services/archive-project.service.ts:
```diff
@@ -37,4 +38,74 @@ export class ArchiveProjectService {
+  }
+
+  showArchiveProjectErrorMessage(archive: boolean): void {
+    this.snackBar.open($localize`Error ${archive ? 'archiving' : 'unarchiving'} unit.`);
```
Make two ```$localize``` statements to make it easier for translators?

``` (pseudocode)
  msg = archive ? $localize`Error archiving unit.` : $localize`Error unarchiving unit.`;
```

hirokiterashima commented on src/app/modules/library/library-project-menu/library-project-menu.component.ts:
```diff
@@ -20,13 +21,15 @@ export class LibraryProjectMenuComponent {
   @Input()
   isRun: boolean;
 
+  archived: boolean = false;
```
Add access modifier?
```suggestion
  protected archived: boolean = false;
```

hirokiterashima commented on src/app/modules/library/library-project/library-project.component.ts:
```diff
@@ -55,4 +65,9 @@ export class LibraryProjectComponent implements OnInit {
     });
   }
+
+  protected projectSelected(event: any): void {
```
Make function name more active?
```suggestion
  protected selectProject(event: any): void {
```

hirokiterashima reviewed (changes requested):
Functionality works well, similar to the run listing page.

I added some code change suggestions inline for improving readability. There were a lot of changes in this PR so I'm not sure if I caught them all, but maybe we can do another round later. Lmk what you think.

geoffreykwan committed:
chore(Unit Library): Clean up archive unit code

#1549

geoffreykwan commented:
@hirokiterashima I made most of the changes and made comments on the ones I left alone.

hirokiterashima commented on src/app/domain/project.ts:
```diff
@@ -91,4 +91,17 @@ export class Project {
     return metadata;
   }
+
+  hasTag(tag: string) {
```
Type return value?
```suggestion
  hasTag(tag: string): boolean {
```

hirokiterashima commented on src/app/domain/project.ts:
```diff
@@ -91,4 +91,17 @@ export class Project {
+    return this.tags.includes(tag);
+  }
+
+  updateArchivedStatus(archived: boolean) {
```
Type return value?
```suggestion
  updateArchivedStatus(archived: boolean): void {
```

hirokiterashima commented on src/app/modules/library/library-project/library-project.component.html:
```diff
@@ -10,7 +10,17 @@
+      class="library-project__checkbox"
+      color="primary"
+      aria-label="Select project"
+      i18n-aria-label
```
Add tooltip above the checkbox instead of aria-label?

hirokiterashima commented on src/app/modules/library/library-project/library-project.component.ts:
```diff
@@ -13,8 +21,10 @@ import { flash } from '../../../animations';
+  @Input() checked: boolean = false;
+  @Input() myUnit: boolean = false;
+  @Input() project: LibraryProject = new LibraryProject();
+  @Output() projectSelectionChanged: EventEmitter<any> = new EventEmitter<any>();
```
Rename?
```suggestion
  @Output() projectSelectionEvent: EventEmitter<any> = new EventEmitter<any>();
```

hirokiterashima commented on src/app/modules/library/library-project/library-project.component.ts:
```diff
@@ -55,4 +65,9 @@ export class LibraryProjectComponent implements OnInit {
+
+  protected selectProject(event: any): void {
+    event.stopPropagation();
+    this.projectSelectionChanged.emit({ checked: event.target.checked, project: this.project });
```
Change to ```projectSelectionEvent``` and use "selected" fieldname instead of ```checked``` to match "selection"?
```suggestion
    this.projectSelectionEvent.emit({ selected: event.target.checked, project: this.project });
```

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.ts:
```diff
@@ -1,21 +1,46 @@
 
-  constructor(protected dialog: MatDialog, protected libraryService: LibraryService) {
+  protected numSelectedProjects: Signal<number> = computed(() => this.selectedProjects().length);
+  protected projectIdToIsSelected: Signal<any> = computed(() =>
```
Give type to signal?
```suggestion
  protected projectIdToIsSelected: Signal<{ [key: number]: boolean }> = computed(() =>
```

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.ts:
```diff
@@ -80,6 +115,69 @@ export class PersonalLibraryComponent extends LibraryComponent {
+  }
+
+  protected projectSelectionChanged(event: any): void {
+    if (event.checked) {
```
Change field name from "checked" to "selected" to match "selection"?
```suggestion
    if (event.selected) {
```

hirokiterashima commented on src/app/modules/library/select-all-projects-checkbox/select-all-projects-checkbox.component.ts:
```diff
@@ -0,0 +1,18 @@
+@Component({
+  selector: 'select-all-projects-checkbox',
+  templateUrl: './select-all-projects-checkbox.component.html',
+  styleUrls: ['./select-all-projects-checkbox.component.scss'],
```
Remove line and the empty scss file?
```suggestion
```

hirokiterashima commented on src/app/modules/library/select-all-projects-checkbox/select-all-projects-checkbox.component.ts:
```diff
@@ -0,0 +1,18 @@
+  @Input() selectedSomeProjects: boolean = false;
+  @Output() selectAllProjectsEvent: EventEmitter<boolean> = new EventEmitter<boolean>();
+
+  protected selectAllProjectsCheckboxClicked(): void {
```
Do we need the function name to contain all projects and checkbox and clicked, or is "select()" clear enough because of the context?
```suggestion
  protected select(): void {
```

hirokiterashima commented on src/app/modules/library/select-all-projects-checkbox/select-all-projects-checkbox.component.html:
```diff
@@ -0,0 +1,8 @@
+  color="primary"
+  [checked]="selectedAllProjects"
+  [indeterminate]="selectedSomeProjects"
+  (click)="selectAllProjectsCheckboxClicked()"
```
Do we need the function name to contain all projects and checkbox and clicked, or is "select()" clear enough because of the context?
```suggestion
  (click)="select()"
```

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.ts:
```diff
@@ -80,6 +115,69 @@ export class PersonalLibraryComponent extends LibraryComponent {
+    this.unselectAllProjects();
+  }
+
+  protected projectSelectionChanged(event: any): void {
```
Make function name more active, and define ProjectSelectionEvent interface?
```suggestion
  protected updateSelectedProjects(event: ProjectSelectionEvent): void {
```

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.ts:
```diff
@@ -80,6 +115,69 @@ export class PersonalLibraryComponent extends LibraryComponent {
+  }
+
+  protected selectAllProjects(selectAll: boolean): void {
+    this.selectedProjects.update(() => (selectAll ? this.getProjectsInView() : []));
```
Use set() here? update() is used when you want to do something with the current value of selectedProjects.
```suggestion
    this.selectedProjects.set(() => (selectAll ? this.getProjectsInView() : []));
```

hirokiterashima reviewed (changes requested):
Great. It's looking better. I added more code improvement suggestions inline. Lmk what you think.

Also, regarding ```unselectAllProjects()```: I don't see the benefit of adding an extra layer of abstraction (i.e., complexity) here, because it's a private function, and its body is just one line. Usually a function is used to group multiple lines of related code together and hide those details from the caller. Who are we abstracting the implementation details from? I think a function like this would make sense if it needs to be called from another class.

Instead, I'm wondering if we can replace calls to ```unselectAllProjects()``` with ```this.selectedProjects.set([])```? ```selectedProjects``` is a class variable and there's no need to call another function to access/set it. We can call it directly, and it's clear what's going on (e.g., you're clearing it by setting it to an empty array) when we read the line- there's no need to look at another function. What do you think?

geoffreykwan committed:
refactor(Unit Library): Clean up archive unit code

#1549

geoffreykwan commented:
I think ```unselectAllProjects()``` is fine. It makes the code easier to read and understand. The function name quickly tells you what is happening. ```this.selectedProjects.set([])``` on the other hand requires more processing to understand what is going on. Using a function also reduces duplication. If there is any change in what needs to be performed when unselecting all projects, you can easily do it in the ```unselectAllProjects()``` function which makes it also easier to change.

hirokiterashima commented on src/app/modules/library/select-all-projects-checkbox/select-all-projects-checkbox.component.ts:
```diff
@@ -0,0 +1,17 @@
+  templateUrl: './select-all-projects-checkbox.component.html',
+  providers: [{ provide: MAT_CHECKBOX_DEFAULT_OPTIONS, useValue: { clickAction: 'noop' } }]
+})
+export class SelectAllProjectsCheckboxComponent {
```
It looks like this component has nothing to do with Projects anymore. Should we rename it to ```SelectAllItemsCheckboxComponent```?
```suggestion
export class SelectAllItemsCheckboxComponent {
```

hirokiterashima commented on src/app/modules/library/select-all-projects-checkbox/select-all-projects-checkbox.component.ts:
```diff
@@ -0,0 +1,17 @@
+  providers: [{ provide: MAT_CHECKBOX_DEFAULT_OPTIONS, useValue: { clickAction: 'noop' } }]
+})
+export class SelectAllProjectsCheckboxComponent {
+  @Input() selectedAllProjects: boolean = false;
+  @Input() selectedSomeProjects: boolean = false;
+  @Output() selectAllProjectsEvent: EventEmitter<boolean> = new EventEmitter<boolean>();
+
+  protected select(): void {
+    this.selectAllProjectsEvent.emit(!(this.selectedAllProjects || this.selectedSomeProjects));
+  }
```
Replace "Project" with "Items"?
```suggestion
  @Input() selectedAllItems: boolean = false;
  @Input() selectedSomeItems: boolean = false;
  @Output() selectAllItemsEvent: EventEmitter<boolean> = new EventEmitter<boolean>();

  protected select(): void {
    this.selectAllItemsEvent.emit(!(this.selectedAllItems || this.selectedSomeItems));
  }
```

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.html:
```diff
@@ -1,7 +1,46 @@
+    <div *ngIf="numSelectedProjects() > 0" class="secondary-text" i18n>
+      {{ numSelectedProjects() }} selected
+    </div>
+    <button
+      mat-icon-button
+      *ngIf="!showArchived && selectedProjects().length > 0"
+      (click)="archiveSelectedProjects(true)"
+      aria-label="Archive selected units"
+      i18n-aria-label
+      matTooltip="Archive"
+      matTooltipPosition="above"
+      i18n-matTooltip
+    >
+      <mat-icon>archive</mat-icon>
+    </button>
+    <button
+      mat-icon-button
+      *ngIf="showArchived && selectedProjects().length > 0"
+      (click)="archiveSelectedProjects(false)"
+      aria-label="Restore selected units"
+      i18n-aria-label
+      matTooltip="Restore"
+      matTooltipPosition="above"
+      i18n-matTooltip
+    >
+      <mat-icon>unarchive</mat-icon>
+    </button>
```
Should we componentize these buttons to ArchiveProjectsButtonComponent? The ArchiveProjectsButtonComponent can show the correct button (archive or unarchive) and also handle the archiving/unarchiving and send the results back to the PersonalLibraryComponent.

Also, we don't need aria-label and i18n-aria-label for these buttons because we already have the matTooltip.

geoffreykwan replied:
Yea we should do this in another issue and then use the component in the unit library and teacher run list.

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.ts:
```diff
@@ -80,6 +116,69 @@ export class PersonalLibraryComponent extends LibraryComponent {
+  public filterUpdated(filterValues: ProjectFilterValues = null): void {
+    super.filterUpdated(filterValues);
+    this.filteredProjects = this.filteredProjects.filter(
+      (project) => project.tags.includes('archived') == this.showArchived
```
Use hasTag()?
```suggestion
      (project) => project.hasTag('archived') == this.showArchived
```

hirokiterashima commented on src/app/modules/library/personal-library/personal-library.component.ts:
```diff
@@ -1,21 +1,47 @@
   filteredProjects: LibraryProject[] = [];
-  personalProjects: LibraryProject[] = [];
-  sharedProjects: LibraryProject[] = [];
+  protected personalProjects: LibraryProject[] = [];
+  projects: LibraryProject[] = [];
+  protected selectedProjects: WritableSignal<LibraryProject[]> = signal([]);
+  protected sharedProjects: LibraryProject[] = [];
+  protected showArchived: boolean = false;
 
-  constructor(protected dialog: MatDialog, protected libraryService: LibraryService) {
+  protected numSelectedProjects: Signal<number> = computed(() => this.selectedProjects().length);
+  protected projectIdToIsSelected: Signal<{ [key: number]: boolean }> = computed(() =>
+    this.selectedProjects().reduce((accumulator, project) => {
+      accumulator[project.id] = true;
+      return accumulator;
+    }, {})
+  );
+  protected selectedAllProjects: Signal<boolean> = computed(() => {
+    const numSelectedProjects = this.numSelectedProjects();
+    return numSelectedProjects !== 0 && numSelectedProjects === this.getProjectsInView().length;
+  });
+  protected selectedSomeProjects: Signal<boolean> = computed(
+    () => this.numSelectedProjects() !== 0 && !this.selectedAllProjects()
+  );
```
Alphabetize field names?

hirokiterashima reviewed (changes requested):
I think we can componentize the archive/unarchive buttons in the PersonalLibraryComponent so the PersonalLibraryComponent doesn't have to handle archiving/unarchiving. We may want to componentize the active/archived selection too, since this component can be reused in the run listing view in the future. I also added a few more code improvement suggestions inline. Lmk what you think.

breity committed:
Move view select to own row, clean up styles

geoffreykwan committed:
refactor(Unit Library): Clean up archive unit code

Rename SelectAllProjectsCheckbox to SelectAllItemsCheckbox.

#1549

hirokiterashima reviewed (commented):
I made a branch https://github.com/WISE-Community/WISE-Client/pull/1567 to refactor LibraryProjectComponent so it doesn't have to calculate if all/some projects are selected. I think that calculation should be done in the SelectAllItemsCheckboxComponent, where it's actually being used. 

Also, I think it'd be clearer to use string literal union keys ('none' | 'some' | 'all') to indicate the current status of the SelectAllItemsCheckboxComponent.

I added more notes in that PR branch.

What do you think?

geoffreykwan commented:
Yea that seems like a good change.

hirokiterashima committed:
refactor(Archive Units): use string union keys (#1567)

* - Use SelectAllItemsStatus string literal union keys ('none' | 'some' | 'all') to indicate the current status of the
SelectAllItemsCheckboxComponent.
- Move status calculation from PersonalLibraryComponent to the SelectAllItemsCheckboxComponent. This requires passing in numSelectedItems and numAllItems.
- Call PersonalLibraryComponent.(un)selectAllProjects() directly as a result of the output from the SelectAllItemsCheckboxComponent

hirokiterashima commented:
I think this looks good enough for now. It's not really ideal to introduce the duplication of archive/unarchive buttons and having ArchiveProjectService dependency in PersonalLibraryComponent, so we should address that sooner than later. Can you create an issue?

There were so many changes in this code, which made the review process difficult. Let's try to break it up into smaller PR's in the future.

geoffreykwan commented:
:tada: This PR is included in version 5.129.0 :tada:

The release is available on [GitHub release](https://github.com/WISE-Community/WISE-Client/releases/tag/v5.129.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
